GUIDELINES.md

```diff
@@ -0,0 +1,153 @@
+# Technical and working style and guidelines
+
+*Challenging this documentation is encouraged. Please set up an issue for challenges, additions or corrections. For explanations, please consult the `dev` Slack channel*
+
+In this document we set out some basic "tips" (either mandatory or encouraged) which should guide your way of working, regardless of the task.
+
+* [Foreword](#foreword)
+* [`python is not pandas`](#python-is-not-pandas)
+* [Design patterns](#design-patterns)
+* [Programming](#programming)
+* [Critical thinking](#critical-thinking)
+* [Naming conventions](#naming-conventions)
+* [Spaces and spacing](#spaces-and-spacing)
+* [Comments and docs](#comments-and-docs)
+
+## Foreword
+
+In advance, we recommend installing the autoformatter [black](https://pypi.org/project/black/) in your IDE. In future we will use `flake8` (or similar) for automatically checking our python codebases. 
+
+Hopefully the conventions laid out here are the easy and intuitive set of [pep8](https://www.python.org/dev/peps/pep-0008/). 
+
+Code reviewers: it is on you to ensure that this style guide has been followed: there are no points for being lenient, but there [non-redeemable] points for being opinionated! We should all feel pressured into making sure that our code meets an acceptable standard.
+
+## `python is not pandas`
+
+tldr; Avoid using `pandas` except for where it makes things ridiculously simple. That is not so say that you should not use `pandas`, but rather that you should generally assume that `pandas` is making your life harder.
+
+Code containing lots of `pandas` operations are almost impossible to review, and therefore have the capacity to accululate vast numbers of bugs.
+
+* In general, `pandas` makes column-wise operations and IO (reading/writing files) dead easy. That said, `pandas` column-wise operations are inherited from `numpy`, and ``numpy` is generally accepted in the place of dataframes.
+* `pandas` is enormous, in many ways. If it can be omitted from your code then you can make big savings in terms of memory usage and requirements clashes, and even CPU time.
+* Instead of Googling how to achieve something in `pandas` with an obscure chaining of functions, break the problem down and solve it yourself. It is highly unlikely that the `pandas` approach to reshaping your data will beat using tools from `numpy`, `itertools`, `functools` and `toolz`, even if you switch to representing data in `numpy` arrays or even as `list` of `dict` (`[{'value': 21}, {'value': 45}]`).
```

Comment:
Agreed it will cause a stir, though it's just a good point to have floating around I think to get people to challenge their assumptions. Should be noted that this isn't an obligatory style point

Perhaps I can change the tldr to:

> Using `pandas` as a means to perform transformations or calculations on your data should be avoided, unless it clearly simplifies the logic and readability of your code. That is not so say that you should not use `pandas`, but rather that you justify to yourself that `pandas` isn't making your life harder in lieu of using standard `python` tools.

@georgerichardson what do you think?

Reply:
(note for George, that the above snippet is `Outdated`)

Reply:
In terms of how to implement it I think we need to be clear what is "Must" which will need clear enforcement procedures and what is "Should" which will be lighter touch. Reviewers will need to be on the same page on this. How difficult do you think people will find this and how much time will it add to their work in the short term?

Reply:
This is all non-controversial / industry standards, so my position is that we sell these as "musts" and then make exceptions, since "shoulds" make reviews too flimsy. 

This anyway overlooks the counterintuitive point that it is *less difficult* for people to work within standard guidelines than by following their own intuition: that is to say that people will develop more robust solutions faster.

Reply:
Cool, I'm all for industry standards! Good to understand the potential short term downsides / pushbacks which it sounds like can be out weighed by the benefits.